server/server.go

```diff
@@ -6,53 +6,287 @@ package server
+	publisher Publisher
+	cfg       ShipperServerConfig
+
+	uuid           string
+	persistedIndex uint64
 
-	queue *queue.Queue
+	polling       polling
+	notifications notifications
+
+	close *sync.Once
 
 	pb.UnimplementedProducerServer
 }
 
-// PublishEvents is the server implementation of the gRPC PublishEvents call
-func (serv shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
-	results := []*messages.EventResult{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event %s: %#v", evt.EventId, evt.Fields.Data)
-		_, err := serv.queue.Publish(evt)
-		if err != nil {
-			// If we couldn't accept any events, return the error directly. Otherwise,
-			// just return success on however many events we were able to handle.
-			if len(results) == 0 {
-				return nil, err
-			}
-			break
+type polling struct {
+	ctx     context.Context
+	stop    func()
+	stopped chan struct{}
+}
+
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(publisher Publisher, cfg ShipperServerConfig) (ShipperServer, error) {
+	if publisher == nil {
+		return nil, errors.New("publisher cannot be nil")
+	}
+
+	err := cfg.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:      id.String(),
+		logger:    logp.NewLogger("shipper-server"),
+		publisher: publisher,
+		cfg:       cfg,
+		polling: polling{
+			stopped: make(chan struct{}),
+		},
+		notifications: notifications{
+			subscribers: make(map[uuid.UUID]chan change),
+			mutex:       &sync.Mutex{},
+		},
+		close: &sync.Once{},
+	}
+
+	s.polling.ctx, s.polling.stop = context.WithCancel(context.Background())
+	s.startPolling()
+
+	return &s, nil
+}
+
+// GetAcceptedIndex atomically reads the accepted index
+func (serv *shipperServer) GetAcceptedIndex() uint64 {
+	return uint64(serv.publisher.AcceptedIndex())
+}
+
+// GetPersistedIndex atomically reads the persisted index
+func (serv *shipperServer) GetPersistedIndex() uint64 {
+	return atomic.LoadUint64(&serv.persistedIndex)
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.
+			With(
+				"expected", serv.uuid,
+				"actual", req.Uuid,
+			).
+			Debugf("shipper UUID does not match, all events rejected")
+
+		return resp, nil
+	}
+
+	for _, e := range req.Events {
+		_, err := serv.publisher.Publish(e)
+		if err == nil {
+			resp.AcceptedCount++
+			continue
+		}
+
+		log := serv.logger.
```

Comment:
Do we need to reconstruct this in the for loop every time for every event? Especially if `With` allocates anything, this seems unnecessary if it is only ever used in debug level logging. It is on the primary data path of the shipper so it will show up in our CPU profiles.

Reply:
It seems like we could just log these in the individual debug log calls so this doesn't cost us anything if debug logging isn't turned on.

Reply:
https://github.com/elastic/elastic-agent-shipper/pull/76/commits/3444327845dc2c8425cc7452e30d90309173a125